### PR TITLE
Visulisation and Visulisation Utilites

### PR DESCRIPTION
**Pull request by DanielPDWalker**
Plotly chart functions, relying on the until.py file for getting dates out of the passed dataframes. These dates are then used in the titles and legends of the plots.

User specified colors are built into the plotting functions but no utility function for them yet. (The plots have a default color palette to use when no colors are passed to them).

Examples of the plots:
groupedbar
linechart
piechart

### CODE REVIEW

**Comment by AKuederle** [comment]
The charts look awesome!

Some general commands:
- Usually when you use git during your devlopment, you would break off you r changes into many small commits instead of one large blob. Just something to keep in mind for the future.
- Also, I think we have a line ending issue.  Have a look that your editor and git are correctly configured to not mark lineending changes as changes (setup.py is shown as complete rewrite even though you just changed a single line.

Specific things to change:
- Check the output of the py35 env CI task. It lists a bunch of code style issues, you should resolve before we can merge. See the contribution guidelines for that.
- It would be nice if we could have unittests for the utils functions
- See specfic comments in code

File: moneypy/visual/utils.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        new_list.append((df['Date'].min(),df['Date'].max()))
+    return new_list
+
+def min_max_date_format(df_list):
+    new_list = []
+    for df in df_list:
+        new_list.append((df['Date'].min().strftime('%y/%m/%d'),df['Date'].max().strftime('%y/%m/%d')))
+    return new_list
+
+def max_date_range(date_tuple_list, date_format=0):
```
Comment (AKuederle) [inline moneypy/visual/utils.py]: Don't use 0 and 1 here. Choose a name that makes clear what the effect is. Also why does a max_date_range function return a formated string? It should return a timedelta.

Comment (AKuederle) [reply moneypy/visual/utils.py]: or at least the start and end as datetime objects

File: moneypy/visual/utils.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    min_date, max_date = date_tuple_list[0][0], date_tuple_list[0][0]
+    for t in date_tuple_list:
```
Comment (AKuederle) [inline moneypy/visual/utils.py]: This whole thing could be significantly less complicated.
Something like this should do the same:
mins, maxs = zip(*data_tuple_list)
return min(mins), max(maxs)

Comment (AKuederle) [reply moneypy/visual/utils.py]: Also current implimentation has a bug. 
```py
min_date, max_date = date_tuple_list[0][0], date_tuple_list[0][0]
```
max_date is set to the min date!

File: moneypy/visual/visulisation.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
+import pandas as pd
+import plotly.graph_objs as go
+import colorlover as cl
+import utils as utils
+
+from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
+
+
+def plot_line_chart(df_list, colors=None):
+    if colors == None:
```
Comment (AKuederle) [inline moneypy/visual/visulisation.py]: This is fine, but could be written as:
```py
colors = colors or cl.scales['10']['div']
```
In general comparisons with `None` are written as `if value is None` or simply `if not value` if you want to allow other values that evalute to False as well (e.g. empty strings, empty lists, False)

File: moneypy/visual/visulisation.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    date_range_tuple = utils.min_max_date(df_list)
+    max_range_tuple = utils.max_date_range(date_range_tuple)
+    date_range_tuple = utils.min_max_date_format(df_list)
+    
+    for index, df in enumerate(df_list):
+        line_chart = go.Scatter(
+            x = df['Date'],
+            y = df['Amount'],
+            hoverinfo='y+x',
+            name = df.name + " (" + str(date_range_tuple[index][0]) + " - " + str(date_range_tuple[index][1]) + ")",
```
Comment (AKuederle) [inline moneypy/visual/visulisation.py]: use `format`

File: moneypy/visual/visulisation.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            tickwidth=2,
+            title = 'Date'),
+        yaxis = dict(title = 'Amount'))
+    
+    fig = dict(data=data, layout=layout)
+    iplot(fig)
+
+
+def plot_grouped_bar_chart(df_list, colors=None):
+    if colors == None:
```
Comment (AKuederle) [inline moneypy/visual/visulisation.py]: see above

File: setup.py [hunk 6]
```diff
@@ -1,22 +1,28 @@
-"""Installation instruction for the MoneyPy module."""
```
Comment (AKuederle) [inline setup.py]: Here it seems like the line ending issue happend

File: requirements.txt [hunk 7]
```diff
@@ -1,2 +1,7 @@
-tox
```
Comment (AKuederle) [inline requirements.txt]: Here it seems like the line ending issue happend

Comment (AKuederle) [inline requirements.txt]: Line ending issue also here

File: moneypy/visual/utils.py (hunk 1 above)
Comment (DanielPDWalker) [reply moneypy/visual/utils.py]: We don't use the dates we get from these functions for anything other than putting them into the title or legend of the plots. Other than min_max_date() which is used for getting the list of dates to compare in the max_date_range() function.

File: moneypy/visual/utils.py (hunk 2 above)
Comment (DanielPDWalker) [reply moneypy/visual/utils.py]: Will look into zip.
``min_date, max_date = date_tuple_list[0][0], date_tuple_list[0][0]`` This shouldn't be a bug, just a way of initializing min_date and max_date to have any date in them so we can compare against them in the for loop. Might just be clearer to put a fixed date in there?

Comment (AKuederle) [reply moneypy/visual/utils.py]: Ah ok... Yes would for sure be clearer. This way feels like a hack

File: moneypy/visual/utils.py (hunk 1 above)
Comment (AKuederle) [reply moneypy/visual/utils.py]: Still a function should do one thing and one thing only. based on the function name, I would assume I get back a datetime object and not a string. The objective of the function is to get the largest range, not to format datetime.

Comment (DanielPDWalker) [reply moneypy/visual/utils.py]: Put the string formatting for dates in its own function, and added a much better default parameter.

File: moneypy/visual/utils.py [hunk 8]
```diff
@@ -12,16 +12,22 @@ def min_max_date_format(df_list):
...
     for t in date_tuple_list:
         for i in t:
             if i < min_date:
                 min_date = i
             elif i > max_date:
                 max_date = i
-    if date_format == 0:
+    return min_date, max_date
+
+def format_date_to_string(date_tuple, date_format="yymmdd")
```
Comment (AKuederle) [inline moneypy/visual/utils.py]: Nice! If you already allow yymmdd or similar as input, should we not just allow a full format string like %y/%m/%d?
Also: 
```
min_date, max_date = date_tuple[0]. date_tuple[1]
``` 
The dot should be a comma...

Comment (DanielPDWalker) [reply moneypy/visual/utils.py]: Yea I noticed the dot, should be changed now. 

We could just let users pass the string format they want, but I was thinking of keeping to a few formats, as some are a lot better than others for display. Otherwise just changing the current format function to take "%y/%m/%d" or "%y/%m" just seems like it will be harder to check is correct / harder for any users with less programming skill. Guess it depends on if we are designing this for  mainly programmers or for as many users as possible.

File: requirements.txt (hunk 7 above)
Comment (DanielPDWalker) [reply requirements.txt]: Looked into this, something to do with changing from a windows file to unix file? The github post I found was for 2014 though so maybe something has changed, Ill keep looking.

**Comment by AKuederle** [comment]
Yes it has something to do with the windows Unix switch, but you can
configure git to handle this

File: requirements.txt (hunk 7 above)
Comment (DanielPDWalker) [reply requirements.txt]: Ive changed a my git global config to core.autocrlf true but it doesnt seem to make a difference.

**Comment by AKuederle** [comment]
Mmh I think changing all lien endings back would be annoying. Just leave it like that, but keep in  mind to check diffs before commiting in the future. If you see changes in every line, even in the ones you have not edited check your line ending settings

**Comment by AKuederle** [comment]
Awesome! Btw. say something if the linter is too anoying. I really like it for my projects, but if it just blocks progress I will disable it

**Comment by DanielPDWalker** [comment]
First I need to look into tox and work out what I'm doing wrong. From my quick googling it seems there is just an issue with python 3.5, but I must be able to do something.

**Comment by AKuederle** [comment]
The linter is also running in the CI. So you can just look at the Travis output to see whats going on. The reason, why only the Py35 pipline fails, is because only this one rund the linter (we don't have to run the linter for every Python version). The other ones just execute the unittests

File: moneypy/visual/visulisation.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
+"""visulisation.py module - Contains the plotting functions.
+
+It uses the utility functions from utils.py,
+to get dates for use in titles and legends of the plots.
+"""
+import pandas as pd
```
Comment (AKuederle) [inline moneypy/visual/visulisation.py]: I think you can just remove it completely. If you don't call pandas, you don't need to import it, even if you are working on pandas dataframes

Comment (DanielPDWalker) [reply moneypy/visual/visulisation.py]: Oh ok, Ill do that.

Comment (AKuederle) [reply moneypy/visual/visulisation.py]: Hahaha why did you push a merge error?

File: tests/visual/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import unittest
+import random
+import pandas as pd
+from moneypy.visual import utils
+
+
+# Set up two empty dataframes.
+df_one = pd.DataFrame()
```
Comment (AKuederle) [inline tests/visual/test_utils.py]: Awesome!

Small suggestion for improvements: the generation of the dfs has a lot of redundant code. How about a `generate_test_df` function? I would also move the generation inside the setUp method in the longterm, to ensure that a testcase does not exidantly mutate the test df and a new one is generated for each testcase

**Comment by AKuederle** [comment]
Mmh there seems to be some issue with the py34 test thing... I am to lazy to debug it... I guess wie are dropping py 3.4 support :D

**Comment by DanielPDWalker** [comment]
Cleaned up the dataframe creation in the unittesting for utils. Also added a few edge case tests.

File: tests/visual/test_utils.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    """Unittest for utils.py"""
+
+    def test_min_max_date(self):
+        # Expected result is a list of tuples of dates
+        self.assertEqual(utils.min_max_date(df_list),
+                         [(df_one['Date'].min(), df_one['Date'].max()),
+                          (df_two['Date'].min(), df_two['Date'].max())])
+        # Edge case of only one dataframe in the list passed.
+        self.assertEqual(utils.min_max_date(df_list_of_one),
```
Comment (AKuederle) [inline tests/visual/test_utils.py]: Just as notes for furture projects. Each of your 3 asserts should be its seperate testcase. Mentality is one test tests exactly on thing. You can be felxible on that rule, but keep in mind, that the rule exists :)

File: tests/visual/test_utils.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    df = pd.DataFrame()
+    df['Date'] = pd.date_range(start=str(date),
+                               periods=number_of_dates,
+                               freq=str(random.randrange(10, 100)) + 'D')
+    df = df.sample(frac=1)
+    return df
+
+
+# Set a random date
+random_date = 2000 + random.randrange(0, 16)
```
Comment (AKuederle) [inline tests/visual/test_utils.py]: I really think we should move that in to the setUp method. Otherwise awesome job on the cleanup

Comment (AKuederle) [reply tests/visual/test_utils.py]: I think this is the only one we should change. Then we just merge it :) We can always come back and change stuff

Comment (DanielPDWalker) [reply tests/visual/test_utils.py]: Moved all the test data generation into setUp in the class.

**Comment by AKuederle** [comment]
Anything you want to change? Otherwise I would merge this

**Comment by DanielPDWalker** [comment]
If your happy with it, I'm happy with it. Like you said we can always come back at a later date and make adjustments.
